broker/src/main/java/org/apache/rocketmq/broker/processor/PopReviveService.java

```diff
@@ -0,0 +1,470 @@
+                if (deliverTime > endTime) {
+                    endTime = deliverTime;
+                }
+            }
+            offset = offset + messageExts.size();
+        }
+        consumeReviveObj.endTime = endTime;
+    }
+
+    private void mergeAndRevive(ConsumeReviveObj consumeReviveObj) throws Throwable {
+        ArrayList<PopCheckPoint> sortList = consumeReviveObj.genSortList();
+        POP_LOGGER.info("reviveQueueId={},ck listSize={}", queueId, sortList.size());
+        if (sortList.size() != 0) {
+            POP_LOGGER.info("reviveQueueId={}, 1st ck, startOffset={}, reviveOffset={} ; last ck, startOffset={}, reviveOffset={}", queueId, sortList.get(0).getStartOffset(),
+                    sortList.get(0).getReviveOffset(), sortList.get(sortList.size() - 1).getStartOffset(), sortList.get(sortList.size() - 1).getReviveOffset());
+        }
+        long newOffset = consumeReviveObj.oldOffset;
+        for (PopCheckPoint popCheckPoint : sortList) {
+            if (!checkAndSetMaster()) {
+                POP_LOGGER.info("slave skip ck process , revive topic={}, reviveQueueId={}", reviveTopic, queueId);
+                break;
+            }
+            if (consumeReviveObj.endTime - popCheckPoint.getReviveTime() <= (PopAckConstants.ackTimeInterval + PopAckConstants.SECOND)) {
+                break;
+            }
+
+            // check normal topic, skip ck , if normal topic is not exist
+            String normalTopic = KeyBuilder.parseNormalTopic(popCheckPoint.getTopic(), popCheckPoint.getCId());
+            if (brokerController.getTopicConfigManager().selectTopicConfig(normalTopic) == null) {
+                POP_LOGGER.warn("reviveQueueId={},can not get normal topic {} , then continue ", queueId, popCheckPoint.getTopic());
+                newOffset = popCheckPoint.getReviveOffset();
+                continue;
+            }
+            if (null == brokerController.getSubscriptionGroupManager().findSubscriptionGroupConfig(popCheckPoint.getCId())) {
+                POP_LOGGER.warn("reviveQueueId={},can not get cid {} , then continue ", queueId, popCheckPoint.getCId());
+                newOffset = popCheckPoint.getReviveOffset();
+                continue;
+            }
+
+            reviveMsgFromCk(popCheckPoint);
+
+            newOffset = popCheckPoint.getReviveOffset();
+        }
+        if (newOffset > consumeReviveObj.oldOffset) {
+            if (!checkAndSetMaster()) {
+                POP_LOGGER.info("slave skip commit, revive topic={}, reviveQueueId={}", reviveTopic, queueId);
+                return;
+            }
+            brokerController.getConsumerOffsetManager().commitOffset(PopAckConstants.LOCAL_HOST, PopAckConstants.REVIVE_GROUP, reviveTopic, queueId, newOffset);
+        }
+        consumeReviveObj.newOffset = newOffset;
+    }
+
+    private void reviveMsgFromCk(PopCheckPoint popCheckPoint) throws Throwable {
+        for (int j = 0; j < popCheckPoint.getNum(); j++) {
```

Comment:
I'm reading pop consuming recently, and confused about retry message handle.
how to handle retry ck message? I could not find ack message of RETRY topic.
@ayanamist can u help to reply? Thanks~

Reply:
1、The pop mode consumes messages. For each message pulled by the client, the broker will save the CK message and wait for the ACK

2、If the ACK message is not received in time due to client exceptions or timeout due to network reasons, popreviveservice will get the CK message from the receive topic and judge whether there is no ack after timeout. If there is no ack after timeout, the message will be put into the retry topic

![image](https://user-images.githubusercontent.com/14222167/128729678-c5962d21-a0f5-4a5e-b388-70d0f5778b78.png)

3、When the client pulls a message, it will probability pull a message from the retry topic
![image](https://user-images.githubusercontent.com/14222167/128730163-14362248-0f33-472e-b299-a10e8ea61ad0.png)

Reply:
@odbozhou Thanks for your quick reply~

I could not find retry messages at consumer client side.
And when popMessage response return, client will call `processPopResponse` method, and it will set the topic of retry message to real topic. 
<img width="955" alt="截屏2021-08-10 上午12 32 34" src="https://user-images.githubusercontent.com/3734319/128742191-fda8317c-6d0b-4dcf-a819-5f84bec9a8d0.png">

<img width="978" alt="截屏2021-08-10 上午12 38 24" src="https://user-images.githubusercontent.com/3734319/128742228-29db122e-6845-43de-884b-cd10f78b53a3.png">

Reply:
As mentioned in the second point above, the consumer pop method pulls the message, and probabilistically pulls the retry message from the broker. The consumer basically does not need to perceive whether the message is pulled from the retry queue or the message pulled from the real topic.

Reply:
I've figured it out. When consumer call `ackAsync` method, the topic will be reset back through extraInfo.
<img width="829" alt="截屏2021-08-10 上午11 20 24" src="https://user-images.githubusercontent.com/3734319/128803533-b18ff729-5f04-44f2-b43f-94a86ce08ea0.png">